### PR TITLE
Build Ray with Bazel

### PR DESCRIPTION
**Pull request by pcmoritz**
Can be used like so:
```
bazel build //:ray_pkg -c opt
cd bazel-genfiles/ray_pkg
ipython
```
and then `import ray`, `ray.init()` etc.

Note that at the moment you need to make sure that the python dependencies of ray are installed, and also that a separate pyarrow installed to be able to use this.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -35,6 +35,9 @@ matrix:
       - rm -rf build
       - ./.travis/install-bazel.sh
       - bazel build ...
+      # Try to run Ray built by Bazel
+      - cd bazel-genfiles/ray_pkg/
+      - python -c "import ray; ray.init()"
```
Comment (robertnishihara) [inline .travis.yml]: The Java build may also install the Ray Python package (I'm not 100% sure).

File: BUILD.bazel [hunk 2]
```diff
@@ -316,3 +340,140 @@ pyx_library(
...
+        "python/ray/experimental/*.py",
+        "python/ray/internal/*.py",
+        "python/ray/WebUI.ipynb",
+        "python/ray/workers/default_worker.py"
+    ])
+)
+
+genrule(
+    name = "redis",
+    cmd = "set -x && curl -sL \"https://github.com/antirez/redis/archive/5.0.3.tar.gz\" | tar xz --strip-components=1 -C . && make && find . && mv ./src/redis-server $(location redis-server)",
```
Comment (robertnishihara) [inline BUILD.bazel]: Nice, version 5 already? I guess we may as well use the latest Redis version. Though if that causes problems we can just use the same version as before and upgrade in a separate PR.

Comment (robertnishihara) [inline BUILD.bazel]: It might be cleaner to use the multi-line syntax like below

Comment (robertnishihara) [inline BUILD.bazel]: What is `location`? Is that bazel specific?

File: BUILD.bazel [hunk 3]
```diff
@@ -316,3 +340,140 @@ pyx_library(
...
+        "//:ray_redis_module",
+        "//:raylet",
+        "//:raylet_monitor",
+        "@plasma//:plasma_store_server"
+    ],
+    outs = ["ray_pkg"],
+    cmd = """
+        set -x &&
+        mkdir -p python/ray &&
+        cp $(locations python/ray/_raylet.so) python/ray &&
```
Comment (robertnishihara) [inline BUILD.bazel]: What does `locations` do?

File: BUILD.bazel [hunk 4]
```diff
@@ -316,3 +340,140 @@ pyx_library(
...
+        "src/ray/gcs/redis_module/ray_redis_module.cc",
+        "src/ray/util/logging.cc"
+    ],
+    deps = [
+      ":ray_common"
+    ]
+)
+
+genrule(
+    name = "pkg",
```
Comment (robertnishihara) [inline BUILD.bazel]: This name is not informative enough

**Comment by robertnishihara** [comment]
@pcmoritz To clarify, Bazel will build all of the relevant stuff, but it will not actually "install" the Python package, right?

We will still need to run `python setup.py install`, which will call Bazel internally, right?

File: BUILD.bazel (hunk 2 above)
Comment (pcmoritz) [reply BUILD.bazel]: Yeah, they added support for streams. Let's try it and see how it goes, given what I know of redis it's probably backwards compatible but let's see when we port all the tests.

Comment (pcmoritz) [reply BUILD.bazel]: Yeah, it just gives the location of the target (it's the recommended way to do it).

File: BUILD.bazel (hunk 3 above)
Comment (pcmoritz) [reply BUILD.bazel]: locations is like location but for multiple locations. Here there is only a single location I think, I changed it.

File: BUILD.bazel (hunk 4 above)
Comment (pcmoritz) [reply BUILD.bazel]: changed

**Comment by pcmoritz** [comment]
Yeah that is correct, we will need to run bazel in `python setup.py install`.

**Review by robertnishihara — APPROVED** [review]

**Comment by robertnishihara** [comment]
cc @rsepassi as a sanity check.

**Review by rsepassi — COMMENTED** [review]
lgtm. maybe in a future pr it would be good to use explicit data dependencies for the cc binaries in the ray py cli. instead of depending on hardcoded paths.

**Pull request merged**